File: selenium/TC-func-REQ-SAF-003-14.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.*;
import org.testng.Assert;
import org.testng.annotations.*;
import java.time.Duration;

public class TC-func-REQ-SAF-003-14_Test {

    @Test
    public void test_TC-func-REQ-SAF-003-14() {
        System.setProperty("webdriver.chrome.driver", "/path/to/chromedriver");
        WebDriver driver = new ChromeDriver();
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(15));
        try {
            // Navigate to the application
            driver.get("http://example.com"); // Replace with actual URL

            // Step 1: Trigger firmware reset
            By resetButton = By.cssSelector("[data-testid='TODO']"); // Firmware reset button
            wait.until(ExpectedConditions.elementToBeClickable(resetButton)).click();

            // Wait for reboot to complete (placeholder for a loading indicator)
            By loadingIndicator = By.cssSelector("[data-testid='TODO']"); // Loading indicator
            wait.until(ExpectedConditions.invisibilityOfElementLocated(loadingIndicator));

            // Verify low fuel warning remains active after reboot
            By lowFuelWarning = By.cssSelector("[data-testid='TODO']"); // Low fuel warning element
            wait.until(ExpectedConditions.visibilityOfElementLocated(lowFuelWarning));
            Assert.assertTrue(driver.findElement(lowFuelWarning).isDisplayed(),
                    "Low fuel warning should remain active after firmware reset");

            // Step 2: Increase fuel level to 121 km
            By fuelInput = By.cssSelector("[data-testid='TODO']"); // Fuel level input field
            WebElement fuelElement = wait.until(ExpectedConditions.elementToBeClickable(fuelInput));
            fuelElement.clear();
            fuelElement.sendKeys("121");

            // Submit the new fuel level
            By submitFuel = By.cssSelector("[data-testid='TODO']"); // Submit button
            wait.until(ExpectedConditions.elementToBeClickable(submitFuel)).click();

            // Verify low fuel warning clears after threshold is exceeded
            wait.until(ExpectedConditions.invisibilityOfElementLocated(lowFuelWarning));
            Assert.assertFalse(driver.findElement(lowFuelWarning).isDisplayed(),
                    "Low fuel warning should clear after fuel level exceeds threshold");

        } finally {
            driver.quit();
        }
    }
}

File: playwright/TC-func-REQ-SAF-003-14.ts
import { test, expect } from '@playwright/test';

test("TC-func-REQ-SAF-003-14 - Verify warning state persistence across firmware reset without auto clearing", async ({ page }) => {
  // Precondition: Low fuel warning active at 110 km, Firmware reset capability enabled
  await page.goto('app://dashboard');

  const lowFuelWarning = page.locator('[data-testid="low-fuel-warning"]');
  await expect(lowFuelWarning).toBeVisible();

  // Step 1: Trigger firmware reset
  const firmwareResetButton = page.locator('[data-testid="firmware-reset-button"]');
  await firmwareResetButton.click();

  await page.waitForLoadState('networkidle');

  // Expected: Low fuel warning remains active after reboot
  await expect(lowFuelWarning).toBeVisible();

  // Step 2: Increase fuel level to 121 km
  const fuelLevelInput = page.locator('[data-testid="fuel-level-input"]');
  await fuelLevelInput.fill('121');

  const updateButton = page.locator('[data-testid="fuel-level-update-button"]');
  await updateButton.click();

  // Expected: Low fuel warning clears
  await expect(lowFuelWarning).toHaveCount(0);
});